File: app/src/main/java/com/vocaby/app/data/dao/VocabyDao.java
package com.vocaby.app.data.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.vocaby.app.data.entity.CustomDefinition;
import com.vocaby.app.data.entity.CustomEntry;
import com.vocaby.app.data.entity.CustomEntryGroup;
import com.vocaby.app.data.entity.EntryWithData;
import com.vocaby.app.data.entity.Type;
import com.vocaby.app.data.entity.User;
import com.vocaby.app.data.entity.UserSaves;
import com.vocaby.app.data.entity.WordDefinitions;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

@Dao
public abstract class VocabyDao {
    @Query("SELECT word FROM dictionary_word UNION " +
            "SELECT entry FROM custom_user_entry " +
            "ORDER BY word COLLATE NOCASE ASC")
    public abstract Single<List<String>> getDictionaryEntries();

    @Query("SELECT EXISTS(SELECT 1 FROM dictionary_word WHERE word = :entry)")
    public abstract Single<Boolean> checkEntryExistence(String entry);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    public abstract Single<Long> createUser(User user);

    @Query("SELECT EXISTS(SELECT * FROM vocaby_user WHERE user_id = :id)")
    public abstract Single<Integer> checkUser(int id);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    public abstract Completable addSave(UserSaves userSaves);

    @Query("DELETE FROM saves WHERE user_id = :id AND word = :word")
    public abstract Completable removeSave(int id, String word);

    @Query("SELECT EXISTS(SELECT 1 FROM saves WHERE word = :word AND user_id = :id)")
    public abstract Single<Integer> hasSave(String word, int id);

    @Query("DELETE FROM saves WHERE user_id = :id")
    public abstract Completable removeAllSaves(int id);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    public abstract Completable addSaves(List<UserSaves> userSaves);

    @Query("SELECT word FROM saves WHERE user_id = :id ORDER BY id DESC")
    public abstract Single<List<String>> getSaves(int id);

    @Transaction
    @Query("SELECT * FROM dictionary_word WHERE word = :word")
    public abstract Single<WordDefinitions> getWordData(String word);

    @Transaction
    @Query("SELECT * FROM dictionary_word WHERE id = :id")
    public abstract Single<WordDefinitions> getWordData(int id);

    @Transaction
    @Query("SELECT * FROM dictionary_word WHERE id = " +
            "(SELECT id FROM dictionary_word ORDER BY RANDOM() LIMIT 1)")
    public abstract Single<WordDefinitions> getRandomWord();

    // CUSTOM ENTRY
    @Insert
    public abstract Single<Long> insertCustomEntry(CustomEntry customEntry);

    @Update
    public abstract Completable updateCustomEntry(CustomEntry customEntry);

    @Insert
    public abstract Single<List<Long>> insertCustomEntryGroups(List<CustomEntryGroup> customEntryGroups);

    @Delete
    public abstract Completable deleteCustomEntryGroups(List<CustomEntryGroup> customEntryGroups);

    @Update
    public abstract Completable updateCustomEntryGroups(List<CustomEntryGroup> customEntryGroups);

    @Insert
    public abstract Completable insertCustomDefinitions(List<CustomDefinition> customDefinitions);

    @Delete
    public abstract Completable deleteCustomDefinitions(List<CustomDefinition> customDefinitions);

    @Update
    public abstract Completable updateCustomDefinitions(List<CustomDefinition> customDefinitions);

    @Query("DELETE FROM custom_user_entry WHERE custom_entry_id = :entryId")
    public abstract Completable deleteUserEntry(int entryId);

    @Query("DELETE FROM custom_user_entry WHERE entry = :entry")
    public abstract Completable deleteUserEntry(String entry);

    @Transaction
    @Query("SELECT * FROM custom_user_entry WHERE user_id = :id AND entry = :entry")
    public abstract Single<EntryWithData> getUserEntryData(int id, String entry);

    @Query("SELECT entry FROM custom_user_entry WHERE user_id = :id ORDER BY last_updated DESC")
    public abstract Single<List<String>> getUserEntries(int id);

    @Query("DELETE FROM custom_user_entry WHERE user_id = :id")
    public abstract Completable clearUserEntries(int id);

    @Insert
    public abstract Completable insertTypes(List<Type> types);

    @Query("SELECT type from entry_type")
    public abstract Single<List<String>> getTypes();
}

File: app/src/main/java/com/vocaby/application/data/dao/VocabyDao.kt
package com.vocaby.application.data.dao

import androidx.room.*
import com.vocaby.application.data.entity.*
import com.vocaby.application.models.customentry.UserEntry
import kotlinx.coroutines.flow.Flow

@Dao
interface VocabyDao {
    /** -------------------- USER -------------------- **/
    @Query("SELECT EXISTS(SELECT * FROM vocaby_user WHERE user_id = :id)")
    suspend fun checkUser(id: Int): Boolean

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun createUser(user: User): Long

    /** --------------------- ENTRY -------------------- **/
    @Delete
    suspend fun deleteEntry(word: Word)

    @Insert
    suspend fun insertEntry(word: Word): Long

    @Insert
    suspend fun insertDefinitions(wordDefinitions: List<Definition>)

    @Query("SELECT EXISTS(SELECT 1 FROM dictionary_word WHERE word = :entry)")
    suspend fun checkEntryExistence(entry: String): Boolean

    @Query(
        "SELECT word FROM dictionary_word WHERE word LIKE :firstLetter||'%' UNION " +
                "SELECT entry FROM custom_user_entry WHERE entry LIKE :firstLetter||'%' " +
                "ORDER BY word ASC"
    )
    suspend fun getDictionaryEntriesByCharacter(firstLetter: String): List<String>

    @Transaction
    @Query("SELECT * FROM dictionary_word WHERE word = :entry")
    suspend fun getEntryData(entry: String): WordDefinitions?

    @Transaction
    @Query("SELECT * FROM dictionary_word WHERE id = :entryId")
    suspend fun getEntryDataWithId(entryId: Int): WordDefinitions?

    /** --------------------- CUSTOM ENTRY -------------------- **/
    @Transaction
    @Query("SELECT * FROM custom_user_entry WHERE user_id = :userId AND entry = :entry")
    suspend fun getUserEntryData(userId: Int, entry: String?): EntryWithData?

    @Transaction
    @Query("SELECT * FROM custom_user_entry WHERE user_id = :userId")
    suspend fun getAllUserEntryData(userId: Int): List<EntryWithData>

    @Transaction
    @Query(
        "SELECT * FROM dictionary_word WHERE id = " +
                "(SELECT id FROM dictionary_word ORDER BY RANDOM() LIMIT 1)"
    )
    suspend fun getRandomWord(): WordDefinitions

    @Query("DELETE FROM custom_user_entry WHERE entry = :entry")
    suspend fun deleteUserEntry(entry: String)

    @Query("DELETE FROM custom_user_entry WHERE custom_entry_id = :entryId")
    suspend fun deleteUserEntry(entryId: Int)

    @Query("SELECT entry, last_updated FROM custom_user_entry WHERE user_id = :id ORDER BY last_updated DESC")
    suspend fun getUserEntries(id: Int): List<UserEntry>

    @Query("DELETE FROM custom_user_entry WHERE user_id = :id")
    suspend fun clearUserEntries(id: Int)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertCustomEntry(customEntry: CustomEntry): Long

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertCustomEntries(customEntries: List<CustomEntry>): List<Long>

    @Update
    suspend fun updateCustomEntry(customEntry: CustomEntry)

    @Insert
    suspend fun insertCustomEntryGroups(customEntryGroups: List<CustomEntryGroup>): List<Long>

    @Delete
    suspend fun deleteCustomEntryGroups(customEntryGroups: List<CustomEntryGroup>)

    @Update
    suspend fun updateCustomEntryGroups(customEntryGroups: List<CustomEntryGroup>)

    @Insert
    suspend fun insertCustomDefinitions(customDefinitions: List<CustomDefinition>)
    @Delete
    suspend fun deleteCustomDefinitions(customDefinitions: List<CustomDefinition>)

    @Update
    suspend fun updateCustomDefinitions(customDefinitions: List<CustomDefinition>)

    /** --------------------- SAVES -------------------- **/
    @Query("SELECT entry FROM saves WHERE user_id = :userId ORDER BY id DESC")
    fun getSavesFlow(userId: Int): Flow<MutableList<String>>

    @Query("SELECT entry FROM saves WHERE user_id = :userId ORDER BY id ASC")
    suspend fun getSaves(userId: Int): List<String>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addSave(userSave: UserSave)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addSaves(userSaves: List<UserSave>)

    @Query("DELETE FROM saves WHERE user_id = :userId AND entry = :entry")
    suspend fun removeSave(userId: Int, entry: String)

    @Query("DELETE FROM saves WHERE user_id = :userId")
    suspend fun clearSaves(userId: Int)

    @Query("SELECT EXISTS(SELECT 1 FROM saves WHERE entry = :entry AND user_id = :userId)")
    fun hasSave(userId: Int, entry: String): Flow<Int>

    /** --------------------- DATA -------------------- **/
    @Insert
    suspend fun recordVisit(dictionaryViewCount: DictionaryViewCount)

    @Insert
    suspend fun recordCustomVisit(customDictionaryViewCount: CustomDictionaryViewCount)

    @Query("SELECT word as entry, COUNT(entry_id) AS count FROM dictionary_view_count " +
            "JOIN dictionary_word ON dictionary_view_count.entry_id = dictionary_word.id GROUP BY entry_id " +
            "UNION " +
            "SELECT entry, COUNT(custom_dictionary_view_count.custom_entry_id) AS count FROM custom_dictionary_view_count " +
            "JOIN custom_user_entry ON custom_dictionary_view_count.custom_entry_id = custom_user_entry.custom_entry_id " +
            "GROUP BY custom_dictionary_view_count.custom_entry_id " +
            "ORDER BY count DESC " +
            "LIMIT :size")
    suspend fun getAllSearchData(size: Int): List<VisitData>

    @Query("SELECT DATE('now', 'start of month')")
    suspend fun getDate(): String


    @Query("SELECT word as entry, COUNT(entry_id) AS count FROM dictionary_view_count " +
            "JOIN dictionary_word ON dictionary_view_count.entry_id = dictionary_word.id " +
            "WHERE DATE(date_visited) BETWEEN DATE('now', 'start of month') AND DATE('now') " +
            "GROUP BY entry_id " +
            "UNION " +
            "SELECT entry, COUNT(custom_dictionary_view_count.custom_entry_id) AS count FROM custom_dictionary_view_count " +
            "JOIN custom_user_entry ON custom_dictionary_view_count.custom_entry_id = custom_user_entry.custom_entry_id " +
            "WHERE DATE(date_visited) BETWEEN DATE('now', 'start of month') AND DATE('now') " +
            "GROUP BY custom_dictionary_view_count.custom_entry_id " +
            "ORDER BY count DESC " +
            "LIMIT :size")
    suspend fun getMonthlySearchData(size: Int): List<VisitData>

    @Query("DELETE FROM dictionary_view_count WHERE user_id = :userId")
    suspend fun deleteVisit(userId: Int)

    @Query("DELETE FROM custom_dictionary_view_count WHERE user_id = :userId")
    suspend fun deleteCustomVisit(userId: Int)

    /** --------------------- TYPES -------------------- **/
    @Insert
    fun insertTypes(vararg types: Type)

    @Query("SELECT type from entry_type")
    suspend fun getTypes(): List<String>
}